packages/realm/src/schema/normalize.ts

```diff
@@ -43,163 +50,430 @@ export const PRIMITIVE_TYPES = new Set<PropertyTypeName>([
-    // Type casting, since it is expected that `normalizePropertyType` moves an object linked type into `objectType`
-    const mergedSchema = {
-      ...removeUndefinedValues(propertySchemaFromType),
-      ...removeUndefinedValues(rest as Omit<ObjectSchemaProperty, "name">),
-    } as ObjectSchemaProperty & { type: PropertyTypeName };
-    if (mergedSchema.type === "mixed" || mergedSchema.objectType === "mixed") {
-      assert(mergedSchema.optional !== false, "Mixed values should be declared as optional");
+  const normalizedSchema =
+    typeof schema === "string"
+      ? normalizePropertySchemaString(name, schema)
+      : normalizePropertySchemaObject(name, schema);
+
+  if (isPrimaryKey) {
+    assert(!normalizedSchema.optional, errMessage(name, "Optional properties cannot be used as a primary key."));
+    normalizedSchema.indexed = true;
+  }
+
+  return normalizedSchema;
+}
+
+/**
+ * Transform and validate a user-provided property schema that is using
+ * the shorthand string notation into its canonical form.
+ */
+function normalizePropertySchemaString(name: ObjectAndPropertyName, schema: string): CanonicalObjectSchemaProperty {
+  assert(schema.length > 0, errMessage(name, "The type must be specified."));
+
+  let type = "";
+  let objectType: string | undefined;
+  let optional = false;
+
+  if (endsWithCollection(schema)) {
+    const end = schema.substring(schema.length - 2);
+    type = COLLECTION_SHORTHAND_TO_NAME[end];
+
+    schema = schema.substring(0, schema.length - 2);
+    assert(schema.length > 0, errMessage(name, `The element type must be specified. (Example: 'int${end}')`));
+
+    const isNestedCollection = endsWithCollection(schema);
+    assert(!isNestedCollection, errMessage(name, "Nested collections are not supported."));
+  }
+
+  if (schema.endsWith("?")) {
+    optional = true;
+
+    schema = schema.substring(0, schema.length - 1);
+    assert(schema.length > 0, errMessage(name, "The type must be specified. (Examples: 'int?', 'int?[]')"));
+
+    const usingOptionalOnCollection = endsWithCollection(schema);
+    assert(
+      !usingOptionalOnCollection,
+      errMessage(
+        name,
+        "Collections cannot be optional. To allow elements of the collection to be optional, use '?' after the element type. (Examples: 'int?[]', 'int?{}', 'int?<>')",
+      ),
+    );
+  }
+
+  if (isPrimitive(schema)) {
+    if (isCollection(type)) {
+      objectType = schema;
+    } else {
+      type = schema as PropertyTypeName;
     }
-    const result: CanonicalObjectSchemaProperty = {
-      indexed: false,
-      optional: false,
-      mapTo: name,
-      ...mergedSchema,
+  } else if (isCollection(schema)) {
+    error(name, "Cannot use the collection name. (Examples: 'int[]' (list), 'int{}' (dictionary), 'int<>' (set))");
+  } else if (schema === "object") {
+    error(name, "To define a relationship, use either 'ObjectName' or { type: 'object', objectType: 'ObjectName' }");
```

Comment:
Should we say `'TypeName'` rather than `'ObjectName'` (ditto below)? My thought is that types have names but objects (aka instances) don't. But maybe realm-js uses "Object" in many places to refer to user defined types? I'll leave this up to the rest of the team since this is just my take as someone who hasn't really used the SDK.

Reply:
The [MongoDB docs](https://www.mongodb.com/docs/realm/sdk/react-native/realm-database/schemas/relationships/) seem to use the wording *Realm object type*. Our parameters when querying object(s) use the word *type* and the JS docs use *The type of Realm object*.

So yeah it's probably a bit more suitable to change this in the user-facing error messages to `'TypeName'`,  `'ObjectTypeName'`, or `'RealmObjectType'`